core/src/main/java/cz/o2/proxima/storage/watermark/GlobalWatermarkTracker.java

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.storage.watermark;
+
+import cz.o2.proxima.annotations.Evolving;
+import cz.o2.proxima.time.WatermarkSupplier;
+import java.time.Instant;
+import java.util.Map;
+
+/**
+ * A tracker of global progress among multiple (distributed) processes.
+ *
+ * <p>A {@link GlobalWatermarkTracker} consists of the following:
+ * <li>Name of the tracker. The name MUST uniquely identify the tracker among all possible global
+ *     trackers. The name also serves as grouping identifier for calculation of the value of global
+ *     watermark.
+ * <li>Predefined list of <i>processes</i>.
+ */
+@Evolving
+public interface GlobalWatermarkTracker extends WatermarkSupplier {
```

Comment:
This makes sense in general 👍 Few concerns here:

- How do we get initial list of consumer?
- How do can we deal with multiple consumers with the same name? eg. multiple partitions consumed in distributed manner. Do we need to know number of consumer instances upfront?

Reply:
We must either know list of initial _process names_ upfront, or be fine with the fact, that tracked watermark can move backwards in time. For the general purposes we intend to use it now, we are fine with watermark moving back in time, so there is actually no reason to configure the process names upfront.

Reply:
**we are fine with watermark moving back in time** 👍 that's correct, this should really simplify usage in beam pipelines for now